src/accelerate/state.py

```diff
@@ -1242,13 +1242,27 @@ def _set_sync_gradients(self, sync_gradients):
 
     def _add_dataloader(self, dataloader):
         "Private function that adds a dataloader to `self.dataloader_references` and sets `in_dataloader` to `True`. Users should not have to call this."
-        self.active_dataloader = dataloader
-        self.dataloader_references.append(self.active_dataloader)
+        self.dataloader_references += [dataloader]
 
     def _remove_dataloader(self, dataloader):
         "Private function that removes a dataloader from `self.dataloader_references` and sets `in_dataloader` to `False` if there are no more dataloaders. Users should not have to call this."
-        self.dataloader_references.remove(dataloader)
-        self.active_dataloader = self.dataloader_references[-1]
+        self.dataloader_references = [
+            dataloader_ref for dataloader_ref in self.dataloader_references if dataloader_ref != dataloader
+        ]
+
+    @property
+    def active_dataloader(self):
+        return self.dataloader_references[-1]
+
+    @property
+    def dataloader_references(self):
+        return [reference() if reference is not None else reference for reference in self._dataloader_references_ref]
```

Comment:
I'm confused, where does `self._dataloader_references_ref` come from?

Reply:
First initializer in the setter, which is not ideal. I'll also add it to `__init__` as I believe that's the proper way to do it with properties.

Reply:
As a heads up, the `self.dataloader_references = [None]` in the `__init__` indirectly initialized the `self._dataloader_references_ref` via the setter.
However, it's cleaner to just initialize the underlying variable directly (e.g. [property docs](https://docs.python.org/3/library/functions.html#property)), which I've done in ca6a5b2c1b6c721ea819d7e0c43cc13941daa24d.